modules/ephemeris/src/main/scala/gem/horizons/EphemerisQuery.scala

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2016-2017 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package gem.horizons
+
+import gem.{ EphemerisKey, Semester }
+import gem.enum.Site
+import gem.math.Ephemeris
+
+import cats.implicits._
+import cats.effect.IO
+
+import fs2.Stream
+import fs2.text.utf8Decode
+
+import org.http4s._
+import org.http4s.client.blaze.PooledHttp1Client
+
+import java.net.URLEncoder
+import java.nio.charset.StandardCharsets.UTF_8
+import java.time.Instant
+import java.time.ZoneOffset.UTC
+import java.time.format.DateTimeFormatter
+import java.util.Locale.US
+
+
+/** Representation of an Horizons ephemeris query.
+  */
+sealed trait EphemerisQuery {
+
+  /** URL string corresponding the the horizons request. */
+  def urlString: String
+
+  /** Makes an horizons request and parses all the ephemeris elements in memory.
+    */
+  def fetchEphemeris: IO[Either[String, Ephemeris]]
+
+  /** Makes an horizons request, parsing and emitting ephemeris elements as they
+    * are received.
+    */
+  def streamEphemeris: Stream[IO, Ephemeris.Element]
+
+}
+
+object EphemerisQuery {
+  type HorizonsDesignation = EphemerisKey.Horizons
+
+  /** Maximum number of elements that may be received by an Horizons ephemeris
+    * request.
+    */
+  val MaxElements: Int = 90024
+
+  // Horizons permits only one request at a time from a given host.
+  private val client = PooledHttp1Client[IO](maxTotalConnections = 1)
+
+  private val Url         = "https://ssd.jpl.nasa.gov/horizons_batch.cgi"
+  private val DateFormat  = DateTimeFormatter.ofPattern("yyyy-MMM-d HH:mm:ss.SSS", US).withZone(UTC)
+
+  private val FixedParams = List(
+    "batch"       -> "1",
+    "CENTER"      -> "coord",
+    "COORD_TYPE"  -> "GEODETIC",
+    "CSV_FORMAT"  -> "NO",
+    "extra_prec"  -> "YES",
+    "MAKE_EPHEM"  -> "YES",
+    "QUANTITIES"  -> "1",
+    "TABLE_TYPE"  -> "OBSERVER",
+    "time_digits" -> "FRACSEC"
+  )
+
+  // Format coordinates and altitude as expected by horizons.
+  private def formatCoords(s: Site): String =
+    f"'${s.longitude.toDoubleDegrees}%1.5f,${s.latitude.toSignedDoubleDegrees}%1.6f,${s.altitude.toDouble/1000.0}%1.3f'"
+
+  // URL encode query params.
+  private def formatQuery(params: List[(String, String)]): String =
+    params
+      .map { case (k, v) => s"$k=${URLEncoder.encode(v, UTF_8.name)}" }
+      .mkString("&")
```

Comment:
I prefer `.intercalate("&")` because `.mkString` calls `.toString` on the elements which can be unsafe. Doesn't matter here since it's strings already, but just FYI.

Reply:
Makes sense, I'll switch it.